notes/gefs_forecasts.qmd

```diff
@@ -0,0 +1,183 @@
+data(station_info, package = 'azmetr')
+
+library(gdalcubes) 
+# remotes::install_github("neon4cast/gefs4cast")
+#remotes::install_github('eco4cast/gefs4cast')
+library(gefs4cast) 
+library(stringr)
+library(lubridate)
+library(stars)
+library(viridisLite)
+library(tmap)
+
+gdalcubes_options(parallel = TRUE)#24)
+```
+
+
+Based on [gefs4cast vignette](https://projects.ecoforecast.org/gefs4cast/articles/gefs-cog.html)
+
+Download ensemble forecasts
+```{r}
+date <- Sys.Date() - lubridate::days(1)
+# may need to adjust for time zones?
+# yesterday's forecast starts today?
+
+gefsdir <- fs::dir_create(file.path("gefs", date))
+
+gefs_cog(gefsdir, 
+         ens_avg = FALSE,
+         max_horizon = 48, # get next 48 h. might only need 24h?          
+         date = date) |>
+  system.time()
+```
+
+```{r}
+gefs_cubes <- list()
+for (i in 1:30) { # ens 00 different?
+  ens <- stringr::str_pad(i, 2, pad = '0')
+  ens_id <- paste0('gep', ens)
+  .f <- fs::dir_ls(gefsdir, 
+                      type="file", 
+                      recurse = TRUE,
+                      regex = ens_id)[-1]
+  step <- .f |> str_extract("f\\d{3}\\.tif") |> str_extract("\\d{3}") 
+  datetime <- Sys.Date() + hours(step) 
+
+  fc <- stars::read_stars(.f[1]) 
+  bandnames <- 
+    stars::st_get_dimension_values(fc, 3) |>
+    stringr::str_extract("([A-Z]+):") |> str_remove(":")
+
+  gefs_cubes[[ens]] <- create_image_collection(.f,
+                                     date_time = datetime, 
+                                     band_names = bandnames)
+}
+```
+
+Create a view that downscales the data to 0.1 degree resolution, hourly.
+```{r}
+
+azmet_pts <- station_info %>% 
+  sf::st_as_sf(coords = c("longitude", "latitude"), crs = 4326)
+  
+azmet_box <- azmet_pts %>% 
+  sf::st_bbox() 
+
+iso <-"%Y-%m-%dT%H:%M:%S"
+ext <- list(t0 = as.character(min(datetime),iso), 
+              t1 = as.character(max(datetime),iso),
+              left = azmet_box[1], right = azmet_box[3],
+              top = azmet_box[4], bottom = azmet_box[2])
+v <- cube_view(srs = "EPSG:4326", 
+               extent = ext,
+              #dx = 0.5, dy = 0.5, # original resolution -- half-degree
+              dx = 0.1, 
+              dy = 0.1,
+              dt = "PT1H",
+              #dt = "PT3H",   # original resolution
+               aggregation = "mean", # will only aggregate if multiple ensemble members are viewed
+               resampling = "cubicspline"
+)
+```
+
+Now generate a dataframe of all the predictions.
+```{r}
+
+preds <- list()
+for(ens in names(gefs_cubes)){
+  gefs_cube <- gefs_cubes[[ens]]
+  x <- raster_cube(gefs_cube, v)  |> 
+         extract_geom(sf = azmet_pts)
+  p <- azmet_pts
+  p$FID <- rownames(azmet_pts)
+  y <- merge(p, x, by = "FID")
+  preds[[ens]] <- y
+}
+ens_forecast <- dplyr::bind_rows(preds, .id = 'ens')
+```
+
+
+Next step: long w/ summaries
+
+```{r}
+## create statistical summaries by station, variable, and time
+
+ens_stats <- ens_long %>% 
```

Comment:
What is `ens_long`?  Looks like a `pivot_longer()` step is missing.  I think probably: 
```suggestion
library(tidyr)
ens_long <- 
  ens_forecast %>% 
  select(-geometry) %>% 
  pivot_longer(APCP:VGRD, names_to = "var")
  
ens_stats <- ens_long %>% 
```